CLDR-17177 kbd: update per feedback (#3495)
unicode-org/cldr: docs/ldml/tr35-keyboards.md

The Unicode Standard and related technologies such as CLDR have dramatically improved the path to language support. However, keyboard support remains platform and vendor specific, causing inconsistencies in implementation as well as timeline.

More and more language communities are determining that digitization is vital to their approach to language preservation and that engagement with Unicode is essential to becoming fully digitized. For many of these communities, however, getting new characters or a new script added to The Unicode Standard is not the end of their journey. The next, often more challenging stage is to get device makers, operating systems, apps and services to implement the script requirements that Unicode has just added to support their language.

> However, commensurate improvements to streamline new language support on the input side have been lacking. CLDR’s new Keyboard Subcommittee has been established to address this very gap.

Reviewer: Waaay overpromising. The keyboard subcommittee cannot "get device makers, operating systems, apps and services to implement the **script requirements** that Unicode has just added to support their language".

Moreover, any mention of the subcommittees / working groups shouldn't be in our specs. (If there other cases we should remove them also.)

![keyboard layout example showing extension of ISO key numbering](images/androidKeyboard.png)

> If it becomes necessary in the future, the format could extend the ISO layout to support keys that are located to the left of the "00" column by using negative column numbers "-01", "-02" and so on, or 100's complement "99", "98",… -->

Reviewer: This is needlessly speculative. No reason to to that. 
```suggestion
If it becomes necessary in the future, the format could extend the ISO layout to support keys that are located to the left of the "00" column. -->
```

Reviewer 2: Note that this is in commented out text which should be removed (or fixed)

Reviewer 2: The format no longer uses ISO layout numbers anyway

**Virtual keyboard:** see **Touch keyboard**

## Notation

- Ellipsis (`…`) in syntax examples are used to denote substituted parts.

>   For example, `id="…keyId"` denotes that `…keyId` (the part between double quotes) is to be replaced with something, in this case a key identifier. As another example, `\u{…}` denotes that the `…` is to be replaced with something, in this case a hex number.

Reviewer 2: ```suggestion
  For example, `id="…keyId"` denotes that `…keyId` (the part between double quotes) is to be replaced with something, in this case a key identifier. As another example, `\u{…usv}` denotes that the `…usv` is to be replaced with something, in this case a Unicode scalar value in hex.
```

I wonder if we should be explicit here -- it's not UTF-16 or other encoding.

Author: good.. should propagate to other sections also since this is just an example

### Escaping

> When explicitly specified, attribute values can contain escaped characters. This specification uses two methods of escaping, the _UnicodeSet_ notation and the `\u{…}` notation.

Reviewer 2: ```suggestion
When explicitly specified, attribute values can contain escaped characters. This specification uses two methods of escaping, the _UnicodeSet_ notation and the `\u{…usv}` notation.
```

### UTS18 Escaping

> The `\u{…}` notation, a subset of hex notation, is described in [UTS #18 section 1.1](https://www.unicode.org/reports/tr18/#Hex_notation). It can refer to one or multiple individual codepoints. Currently, the following attribute values allow the `\u{…}` notation:

Reviewer: why …usv above, and just … here?

Author: fixed

### UTS18 Escaping

> The `\u{…}` notation, a subset of hex notation, is described in [UTS #18 section 1.1](https://www.unicode.org/reports/tr18/#Hex_notation). It can refer to one or multiple individual codepoints. Currently, the following attribute values allow the `\u{…}` notation:

Reviewer 2: ```suggestion
The `\u{…usv}` notation, a subset of hex notation, is described in [UTS #18 section 1.1](https://www.unicode.org/reports/tr18/#Hex_notation). It can refer to one or multiple individual codepoints. Currently, the following attribute values allow the `\u{…usv}` notation:
```

Ambiguous to me if we should do this but perhaps for consistency

## File and Directory Structure

* In the future, new layouts will be included in the CLDR repository, as a way for new layouts to be distributed in a cross-platorm manner. The process for this repository of layouts has not yet been defined, see [CLDR-17253](https://unicode-org.atlassian.net/browse/CLDR-17253) for up-to-date information.  Layouts which are currently in the repository are intended as sample data only.

> 

Reviewer: I could expect to see just sample data in the repository when the spec is in tech preview. Once it is approved, however, I would expect to see real data there (and no sample data).

Author: The challenge is that we haven't defined the process for inclusion in the common repository yet. 

The Bengali and Maltese keyboards are real keyboards, not toys. The French azerty keyboard is a test, that could be withdrawn for now.

Perhaps something else here, saying that the process for inclusion is still being defined, and strike the section about sample data?

Author: I've updated this to remove the mention of sample data. I'll plan to remove the only keyboard that's only sample data, the azerty one.

Author: Mark, I've re-added the DTDs so that sample files pass.  CLDR tooling breaks without a DTD.  Split this out to ticket https://unicode-org.atlassian.net/browse/CLDR-17371

* Layouts have version metadata to indicate their specification compliance versi​​on number. For this tech preview, the value used must be `techpreview`. When the the specification is out of tech preview, a specific CLDR version number will be used. See [`cldrVersion`](tr35-info.md#version-information).

```xml
<keyboard3 conformsTo="techpreview"/>
```

> > _Note_: Unlike other LDML files, layouts are designed to be used outside of the CLDR source tree.  A new mechanism for referencing the DTD path should ideally be used, such as a URN or FPI. See <https://unicode-org.atlassian.net/browse/CLDR-15505> for discussion. For this tech preview, a relative path to the dtd will continue to be used in sample data.  Future versions may give other recommendations.

Reviewer: "A new mechanism" is way too vague. You might say "before the specification exits tech preview, the intent is to have a different mechanism ..."

The sample data shouldn't use a mechanism that you discourage.

Also, I believe that you allow keyboards to import data: there should be a strong warning about pulling an isolated keyboard out of the repository without everything it depends on, including the dtd.

Author: Removed this section and removed DOCTYPEs.  (all tests broken, but will fix)

```xml
<!DOCTYPE keyboard3 SYSTEM "../dtd/ldmlKeyboard3.dtd">
```

> > For distributing keyboards outside of the CLDR repository, it is recommended to either use a copy of the .dtd file, omit the `DOCTYPE` clause. Do not use a http or https URL in the DOCTYPE.

Reviewer 2: ```suggestion
> For distributing keyboards outside of the CLDR repository, it is recommended to either use a copy of the .dtd file, or omit the `DOCTYPE` clause. Do not use a http or https URL in the DOCTYPE.
```

## File and Directory Structure

> * In the future, new layouts will be included in the CLDR repository, as a way for new layouts to be distributed in a cross-platorm manner. The process for this repository of layouts has not yet been defined, see [CLDR-17253](https://unicode-org.atlassian.net/browse/CLDR-17253) for up-to-date information.

Author: I'll change to point to the keyboard WG page

> * In the future, new layouts will be included in the CLDR repository, as a way for new layouts to be distributed in a cross-platorm manner. The process for this repository of layouts has not yet been defined, see [CLDR-17253](https://unicode-org.atlassian.net/browse/CLDR-17253) for up-to-date information.

Author: done

> * Layouts have version metadata to indicate their specification compliance versi​​on number. For this tech preview, the value used must be `techpreview`. When the the specification is out of tech preview, a specific CLDR version number will be used. See [`cldrVersion`](tr35-info.md#version-information).

Reviewer: Very soon, this should change...

Author: yes. ideally this would merge first.

_Attribute:_ `locale` (required)

> This attribute contains the primary locale of the keyboard using BCP 47 [Unicode locale identifiers](tr35.md#Canonical_Unicode_Locale_Identifiers) - for example `"el"` for Greek. Sometimes, the locale may not specify the base language. For example, a Devanagari keyboard for many languages could be specified by BCP-47 code: `"und-Deva"`. However, it is better to list out the languages explicitly using the [`locales`](#element-locales) element.

Reviewer: attribute **value** is

_Attribute:_ `output`

> > The `output` attribute contains the sequence of characters that is emitted when pressing this particular key. Control characters, whitespace (other than the regular space character) and combining marks in this attribute are escaped using the `\u{…}` notation. More than one key may output the same output.

Reviewer 2: ```suggestion
> The `output` attribute contains the sequence of characters that is emitted when pressing this particular key. Control characters, whitespace (other than the regular space character) and combining marks in this attribute are escaped using the `\u{…usv}` notation. More than one key may output the same output.
```

### Element: displays

The `displays` element consists of a list of [`display`](#element-display) subelements.

**Syntax**

```xml
<displays>
    <display ... />
    <display ... />

>     ...

Reviewer 2: Use ellipsis here?

### Element: display

The display elements can be used to describe what is to be displayed on the keytops for various keys. For the most part, such explicit information is unnecessary since the `@to` element from the `keys/key` element will be used for keytop display. However some characters, such as diacritics, do not display well on their own.

Another useful scenario is where there are doubled diacritics, or multiple characters with spacing issues.

Finally, this provides a way to specify the keytop for keys which do not otherwise produce output, which is common for the `@layerId` attribute.

Note: `displays` elements are designed to be shared across many different keyboard layout descriptions, and imported with `<import>` where needed.

For combining characters, U+25CC `◌` is used as a base. It is an error to use a combining character without a base in the `display` attribute.

For example, a key which outputs a combining tilde (U+0303) can be represented as follows:

```xml

>     <display output="\u{0303}" display="◌̃" />  <!-- \u{25CC} \u{0303}-->

Reviewer 2: ```suggestion
    <display output="\u{0303}" display="◌̃" />  <!-- \u{25CC} \u{0303}-->
    <display output="\u{0303}" display="\u{25cc}\u{0303}" />  <!-- also acceptable -->
```

Given how badly it displays in GitHub preview monospace here, the current example may be counter-productive?

![image](https://github.com/unicode-org/cldr/assets/4498365/f8eac13a-73ed-44e2-b6ee-090fa5da361d)

Author: maybe, though it's just an example.

### Element: display

The display elements can be used to describe what is to be displayed on the keytops for various keys. For the most part, such explicit information is unnecessary since the `@to` element from the `keys/key` element will be used for keytop display. However some characters, such as diacritics, do not display well on their own.

> Another useful scenario is where there are doubled diacritics, or multiple characters with spacing issues.

Reviewer: Minor; a list of single line paragraphs is strange. Do you mean this to be a list?

### Element: display

The `display` elements can be used to describe what is to be displayed on the keytops for various keys. For the most part, such explicit information is unnecessary since the `@to` element from the `keys/key` element will be used for keytop display.

- Some characters, such as diacritics, do not display well on their own.
- Another useful scenario is where there are doubled diacritics, or multiple characters with spacing issues.
- Finally, the `display` element provides a way to specify the keytop for keys which do not otherwise produce output. Keys which switch layers using the `@layerId` attribute typically do not produce output.

> Note: `displays` elements are designed to be shared across many different keyboard layout descriptions, and imported with `<import>` where needed.

#### Non-spacing marks on keytops

> For non-spacing marks, U+25CC `◌` is used as a base. It is an error to use a combining character without a base in the `display` attribute. For example, `display="\u{0303}"` would produce an error.

Reviewer: ```suggestion
For non-spacing marks, U+25CC `◌` is used as a base. It is an error to use a nonspacing character without a base in the `display` attribute. For example, `display="\u{0303}"` would produce an error.
```

A key which outputs a combining tilde (U+0303) could be represented as either of the following:

```xml
    <display output="\u{0303}" display="◌̃" />  <!-- \u{25CC} \u{0303}-->
    <display output="\u{0303}" display="\u{25cc}\u{0303}" />  <!-- also acceptable -->
```

This way, a key which outputs a combining tilde (U+0303) will be represented as `◌̃` (a tilde on a dotted circle).

> Some scripts/languages may prefer a different base than U+25CC. See  [`<displayOptions baseCharacter=…/>`](#element-displayoptions).

Reviewer: script and languages never prefer anything; they are not animate.

```suggestion
Users of some scripts/languages may prefer a different base than U+25CC. See  [`<displayOptions baseCharacter=…/>`](#element-displayoptions).
```

_Attribute:_ `value` (required)

> > Strings may contain whitespaces. However, for clarity, it is recommended to escape non-spacing marks, even in strings.

Reviewer 2: I think this was original supposed to refer to whitespace characters rather than "spacing marks" (which is kinda weird).

```suggestion
> Strings may contain whitespaces. However, for clarity, it is recommended to escape whitespace characters, even in strings.
```

3.  No `keyMap[@modifiers]` value can overlap with another `keyMap[@modifiers]` value.
    * eg you can't have `"RAlt Ctrl"` in one `keyMap`, and `"Alt Shift"` in another (because Alt = RAltLAlt).
4.  Every sequence of characters in a `transform[@from]` value must be a concatenation of two or more `map[@to]` values.

>     * eg with `<transform from="xyz" to="q">` there must be some map values to get there, such as `<map… to="xy">` & `<map… to="z">`

Reviewer: This looks like the element has a prefix "map", like "mapStuff". If you really mean intervening attributes, put a space between map and …

Reviewer 2: Also in commented out section that needs to be fixed

Reviewer: Why is this resolved?

Reviewer 2: The entire commented out section was removed.

> 3.  No `keyMap[@modifiers]` value can overlap with another `keyMap[@modifiers]` value.

Reviewer 2: Just realised this section is all commented out and wrong anyway (e.g. there is no `<map` element. It's a TODO that should be resolved before we publish.

Author: I'm going to remove the TODO as part of the general cleanup ticket.
